src/aws/flb_aws_imds.c

```diff
@@ -0,0 +1,340 @@
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <fluent-bit/flb_info.h>
+#include <fluent-bit/flb_sds.h>
+#include <fluent-bit/flb_http_client.h>
+#include <fluent-bit/flb_aws_credentials.h>
+#include <fluent-bit/flb_aws_util.h>
+#include <fluent-bit/flb_jsmn.h>
+#include <fluent-bit/aws/flb_aws_imds.h>
+
+#define FLB_AWS_IMDS_ROOT                                   "/"
+#define FLB_AWS_IMDS_V2_TOKEN_PATH                          "/latest/api/token"
+
+/* Request headers */
+static struct flb_aws_header imds_v2_token_ttl_header = {
+    .key = "X-aws-ec2-metadata-token-ttl-seconds",
+    .key_len = 36,
+    .val = "21600", // 6 hours (ie maximum ttl)
+    .val_len = 5,
+};
+
+/* Request header templates */
+const static struct flb_aws_header imds_v2_token_token_header_template = {
+    .key = "X-aws-ec2-metadata-token",
+    .key_len = 24,
+    .val = "",      // Replace with token value
+    .val_len = 0,   // Replace with token length
+};
+
+/* Declarations */
+static int get_imds_version(struct flb_aws_imds *ctx);
+static int refresh_imds_v2_token(struct flb_aws_imds *ctx);
+
+/* Default config values */
+const struct flb_aws_imds_config flb_aws_imds_config_default = {
+    FLB_AWS_IMDS_VERSION_EVALUATE
+};
+
+/* Create IMDS context */
+struct flb_aws_imds *flb_aws_imds_create(struct flb_config *config,
+                       struct flb_aws_imds_config *imds_config,
+                       struct flb_aws_client *ec2_imds_client)
+{
+    struct flb_aws_imds *ctx = NULL;
+
+    /* Create context */
+    ctx = flb_calloc(1, sizeof(struct flb_aws_imds));
+    if (!ctx) {
+        flb_errno();
+        return NULL;
+    }
+
+    /* 
+     * Set IMDS version to whatever is specified in config
+     * Version may be evaluated later if set to FLB_AWS_IMDS_VERSION_EVALUATE
+     */
+    ctx->imds_version = imds_config->use_imds_version;
+    ctx->imds_v2_token = flb_sds_create_len("INVALID_TOKEN", 13);
+
+    /* Detect IMDS support */
+    if (!ec2_imds_client->upstream) {
+        flb_debug("[imds] unable to connect to EC2 IMDS. ec2_imds_client upstream is null");
+
+        flb_free(ctx);
+        return NULL;
+    }
+    if (0 != strncmp(ec2_imds_client->upstream->tcp_host,
+                            FLB_AWS_IMDS_HOST,
+                            FLB_AWS_IMDS_HOST_LEN)) {
+        flb_debug("[imds] ec2_imds_client tcp host must be set to %s",
+                  FLB_AWS_IMDS_HOST);
+        flb_free(ctx);
+        return NULL;
+    }
+    if (ec2_imds_client->upstream->tcp_port != FLB_AWS_IMDS_PORT) {
+        flb_debug("[imds] ec2_imds_client tcp port must be set to %i",
+                  FLB_AWS_IMDS_PORT);
+        flb_free(ctx);
+        return NULL;
+    }
+    
+    /* Connect client */
+    ctx->ec2_imds_client = ec2_imds_client;
+    return ctx;
+}
+
+/* Destroy IMDS context */
+void flb_aws_imds_destroy(struct flb_aws_imds *ctx) {
+    if (ctx->imds_v2_token) {
```

Comment:
For Fluent Bit general requirements - the opening brace is always in the next line: https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#general-requirements

Reply:
Shucks, good comment. I'll refactor.

Reply:
reminder to do this

Reply:
Should be fixed, no? I just forgot to hit resolved. GitHub might link to a previous commit

Reply:
See: https://github.com/fluent/fluent-bit/pull/4086/files#diff-489ce2c73f3c568d0a9c840c52f25ba1732770cae1ad388e929aa027ab92eed4R99